### PR TITLE
fix: Correct status message for nodes during scale-up

### PR DESCRIPTION
**Pull request by matthias-horne**
**How to categorize this PR?**

/area usability
/kind bug

**What this PR does / why we need it**:

Without this fix the node status was reported as `too many worker nodes are registered. Exceeding maximum desired machine count` under certain conditions while the cluster was still scaling up. 

The implementation was introduced in #11779 with a solution described here: https://github.com/gardener/gardener/pull/11084#issuecomment-2639137008

It has the following problematic edge cases:
- `checkNodesScalingDown` is called unconditionally, even if `registeredNodes < desiredMachines`, leading to the unexpected status message.
- `checkNodesScalingUp` is called only if the condition `MachineDeploymentAvailable` is not true for at least one `MachineDeployment`. However, a `MachineDeployment` can still be scaling up after the condition is true when `maxUnavailable > 0`.
  - MCM sets `MachineDeploymentAvailable` to `true` when `availableReplicas >= Spec.Replicas - maxUnavailable` ([Source](https://github.com/gardener/machine-controller-manager/blob/bd15a0127dce75ee533a6a917e664ffd0bab68c6/pkg/controller/deployment_sync.go#L655-L662))
  - As a result the node status was reporting `too many worker nodes` when `availableReplicas >= Spec.Replicas - maxUnavailable && availableReplicas < Spec.Replicas`.

The new logic ignores the the `MachineDeploymentAvailable` and relies purely on comparing `registeredNodes` with `desiredMachines` with a special case for rolling updates (not sure if this is even required). Tests had to be adjusted slightly because the test data for scale-up tests relied purely on the `MachineDeploymentAvailable` to be false

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```bugfix user
Fix an edge case where node status was reported as "too many worker nodes" while the cluster was still scaling up.
```

### CODE REVIEW

File: pkg/gardenlet/controller/shoot/care/health.go [hunk 1]
```diff
@@ -953,7 +948,7 @@ func checkNodesScalingUp(machineList *machinev1alpha1.MachineList, readyNodes, d
 }
 
 func checkNodesScalingDown(machineList *machinev1alpha1.MachineList, nodeList []*corev1.Node, registeredNodes, desiredMachines int) error {
-	if registeredNodes == desiredMachines {
+	if registeredNodes <= desiredMachines {
```
Comment (plkokanov) [inline pkg/gardenlet/controller/shoot/care/health.go]: This function is only called when `registeredNodes > desiredMachines` so this if check will always be false.

File: pkg/gardenlet/controller/shoot/care/health.go [hunk 2]
```diff
@@ -905,21 +898,23 @@ func CheckNodesScaling(ctx context.Context, seedClient client.Client, nodeList [
 		}
 	}
 
-	if checkScaleUp {
+	if registeredNodes < desiredMachines {
```
Comment (plkokanov) [inline pkg/gardenlet/controller/shoot/care/health.go:901]: I wonder if we should instead use 
```suggestion
	if registeredNodes <= desiredMachines {
```

I think that we have the following `readyAndSchedulableNodes <= registeredNodes <= desiredMachines` and we would want to check for scale up when `readyAndSchedulableNodes < desiredMachines`, which coul also be true when `registeredNodes == desiredMachines`

**Review by plkokanov — COMMENTED** [review]
Thanks for the PR!

Small clarification for this:
>  with a special case for rolling updates (not sure if this is even required)

The proposal in https://github.com/gardener/gardener/pull/11084#issuecomment-2574651958 and the https://github.com/gardener/gardener/pull/11779 and https://github.com/gardener/gardener/pull/11869 PRs attempted to fix the case when there is a rolling update. During a rolling update, the number of nodes could be more than the desired machines, similar to what happens when scaling down, so we wanted to handle it separately and display a different condition reason.

---

The logic that checks the `MachineDeploymentAvailable` condition was present from before that, and I think we didn't want to change it in those PRs. I don't remember anymore, what was the specific reason due to which we kept it, but maybe the idea behind it was that we don't care if scale up is still ongoing when the condition is `True` - e.g. if the scale up has already covered the min availability, we don't report it as a problem even if `readyAndSchedulableNodes < desiredMachines`. I guess we could decide now whether this is intended behavior and document it, or proceed with the removal of this condition check.

In any case, I think you're right that the `checkNodesScalingDown` should only be executed when `registeredNodes > desiredMachines`

File: pkg/gardenlet/controller/shoot/care/health.go (hunk 2 above)
Comment (matthias-horne) [reply pkg/gardenlet/controller/shoot/care/health.go:901]: That's a good catch and I fully agree! 

When changing I had to change one unit test, which was failing. Also added two new unit tests to ensure this behavior will be tested in the future. Removed one redundant unit test. Since all the tests are very similar git found a way to make the diff really confusing. Checking out the PR and using `git diff --histogram` provides a better overview.

File: pkg/gardenlet/controller/shoot/care/health.go (hunk 1 above)
Comment (matthias-horne) [reply pkg/gardenlet/controller/shoot/care/health.go]: I left it in place initially as a matter of defensive programming. Since this is an internal method, I agree that it can be removed.

**Comment by matthias-horne** [comment]
> The logic that checks the `MachineDeploymentAvailable` condition was present from before that, and I think we didn't want to change it in those PRs. I don't remember anymore, what was the specific reason due to which we kept it, but maybe the idea behind it was that we don't care if scale up is still ongoing when the condition is `True` - e.g. if the scale up has already covered the min availability, we don't report it as a problem even if `readyAndSchedulableNodes < desiredMachines`. I guess we could decide now whether this is intended behavior and document it, or proceed with the removal of this condition check.

Here are my arguments in favor of always reporting the scale up: 
1. The number of `desiredMachines` is usually increased by the cluster autoscaler when there are pods that cannot be scheduled on the existing nodes. In this case it is valuable information for the shoot owner to know that new nodes are currently being provisioned.
2. We have seen shoot clusters where `maxUnavailable` is 100%. I guess this is not necessarily a good setting. It might help in the case of scale to 0 for the worker group (haven't checked). With 100% `maxUnavailable` `MachineDeploymentAvailable` will always be true and a scale up will never be reported.

**Comment by plkokanov** [comment]
> Here are my arguments in favor of always reporting the scale up:
> 
> 1. The number of `desiredMachines` is usually increased by the cluster autoscaler when there are pods that cannot be scheduled on the existing nodes. In this case it is valuable information for the shoot owner to know that new nodes are currently being provisioned.
> 2. We have seen shoot clusters where `maxUnavailable` is 100%. I guess this is not necessarily a good setting. It might help in the case of scale to 0 for the worker group (haven't checked). With 100% `maxUnavailable` `MachineDeploymentAvailable` will always be true and a scale up will never be reported.

I agree. The only additional thing I was considering is whether this will cause more noise for gardener operators, but that should already be the case before the PR due to the incorrect fall-through to `checkNodesScalingDown`

**Review by plkokanov — COMMENTED** [review]
Thanks a lot for addressing the comments and going through the suggestions!

/lgtm
/approve

File: pkg/gardenlet/controller/shoot/care/health.go [hunk 3]
```diff
@@ -904,21 +898,29 @@ func CheckNodesScaling(ctx context.Context, seedClient client.Client, nodeList [
...
 		if err := checkNodesScalingUp(machineList, readyAndSchedulableNodes, desiredMachines); err != nil {
 			return "NodesScalingUp", err
 		}
 	}
 
-	if err := checkNodesScalingDown(machineList, nodeList, registeredNodes, desiredMachines); err != nil {
-		return "NodesScalingDown", err
+	if readyAndSchedulableNodes < desiredMachines {
```
Comment (timuthy) [inline pkg/gardenlet/controller/shoot/care/health.go:908]: Can you please add a doc string explaining the circumstances and special case that is checked here?

**Review by timuthy — COMMENTED** [review]
Thanks for the revised version of the nodes check.

File: pkg/gardenlet/controller/shoot/care/health.go (hunk 3 above)
Comment (matthias-horne) [reply pkg/gardenlet/controller/shoot/care/health.go:908]: Comment added

**Review by timuthy — COMMENTED** [review]
/lgtm
/approve
